### PR TITLE
chore: add linter for google style guide-compliant python imports.

### PR DESCRIPTION
**Pull request by ioga**
## Description

I've just been reminded once again that we have to follow google style guide on python imports: https://google.github.io/styleguide/pyguide.html#22-imports

>Use import statements for packages and modules only, not for individual classes or functions.

This PR adds `flake8-import-restrictions` plugin which checks for these violations.

We had 689 violations for this rule in `harness/`. Since I don't have courage to write a script to fix this automatically, I've fixed all violations in the Core API `determined/core/*` and added the remaining violating files to the ignore list. Thus, all new files under `harness/` will be checked.

One more thing this style guide rule disallows is relative imports: `Do not use relative names in imports. Even if the module is in the same package, use the full package name`. This flake8 plugin crashes on `from . import ...`, so I had to fix this in a few places to prevent it from crashing.

The error is fully ignored in `model_hub`, `e2e_tests`, and `examples/tests`.

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by hamidzr — APPROVED** [review]

File: harness/determined/cli/agent.py [hunk 1]
```diff
@@ -1,20 +1,17 @@
...
 import typing
-from collections import OrderedDict
-from operator import attrgetter
 from typing import Any, Callable, Dict, List
 
 import determined.cli.render
 from determined import cli
-from determined.cli import render
+from determined.cli import render, errors
 from determined.cli import task as cli_task
```
Comment (hamidzr) [inline harness/determined/cli/agent.py]: it'd be nice if some tool auto-merged these.

File: harness/determined/cli/agent.py [hunk 2]
```diff
@@ -79,16 +76,16 @@ def list_slots(args: argparse.Namespace) -> None:
         if r["container_id"]
     }
 
-    def device_type_string(deviceType: typing.Optional[devicev1Type]) -> str:
-        if deviceType == devicev1Type.CUDA:
+    def device_type_string(deviceType: typing.Optional[bindings.devicev1Type]) -> str:
```
Comment (hamidzr) [inline harness/determined/cli/agent.py:79]: just a thought: IMO for some very common modules it might make sense to break the rule and use shorter aliases (eg how they do `numpy as np` or pandas as pd) but then it might not be clear to everyone and this loses its value.

`bindings` or `typing` would be good candidates.

File: harness/determined/cli/agent.py (hunk 1 above)
Comment (ioga) [reply harness/determined/cli/agent.py]: it turns out google style guide requires one import per line: https://google.github.io/styleguide/pyguide.html#313-imports-formatting
i.e. 
```
from determined.cli import render
from determined.cli import errors 
```
but we have the formatter setup the opposite way.

File: harness/determined/cli/agent.py (hunk 2 above)
Comment (ioga) [reply harness/determined/cli/agent.py:79]: `typing` is already excepted from this rule. adding our own custom exception for anything else would mean a fork.

bindings is not the worst thing. for example, in CLI `from determined.common import declarative_argparse` looks pretty bad. agreeing on a standardized import alias, e.g. `from determined.common import declarative_argparse as dap` would be nice.

File: e2e_tests/.flake8 [hunk 3]
```diff
@@ -4,9 +4,57 @@ max-line-length = 100
 # We ignore F401 in __init__.py because it is expected for there to be
 # "unused imports" when defining a "regular" package. (This file is
 # implicitly executed when the package is imported, and the imports would
-# be used by the importer.) We ignore patch_saver_restore.py because it includes
-# a near-verbatim TensorFlow function with a small patch.
-per-file-ignores = __init__.py:F401 patch_saver_restore.py:E111,E114,
+# be used by the importer.)
+per-file-ignores =
```
Comment (wes-turner) [inline e2e_tests/.flake8:8]: Do you think this is worth a ticket and/or TODO referencing that ticket?

File: e2e_tests/Makefile [hunk 4]
```diff
@@ -20,7 +20,7 @@ check:
 		-s ./tests/fixtures/pytorch_lightning_amp/mnist.py \
 		-s ./tests/fixtures/pytorch_lightning_amp/data.py
 	black . --check
-	flake8
+	python -m flake8
```
Comment (wes-turner) [inline e2e_tests/Makefile:23]: As you pointed out earlier, this is odd. Maybe worth a comment since it represents a real, engineering decision? Up to you.

File: e2e_tests/tests/nightly/test_convergence.py [hunk 5]
```diff
@@ -161,8 +161,8 @@ def test_unets_tf_keras_accuracy(client: _client.Determined) -> None:
 
 @pytest.mark.nightly
 def test_gbt_titanic_estimator_accuracy(client: _client.Determined) -> None:
-    import tensorflow as tf
-    from packaging import version
+    import tensorflow as tf  # noqa: I2000
+    from packaging import version  # noqa: I2000
```
Comment (wes-turner) [inline e2e_tests/tests/nightly/test_convergence.py:165]: I don't think either of these `noqa`s are used. I _believe_ `version` is a module, and the `as` rule makes sure that the alias is two letters long or more.

File: e2e_tests/.flake8 [hunk 6]
```diff
@@ -4,9 +4,57 @@ max-line-length = 100
...
+  __init__.py:F401,I2041
```
Comment (wes-turner) [inline e2e_tests/.flake8:9]: I don't understand the `I*` ignore directives. They're `Innnn`, where according to the [README](https://github.com/atollk/flake8-import-restrictions) the rules are of the form `IMRnnn`. Why does this work? Are these the aliases we should be using for the rules in question?

File: examples/tests/.flake8 [hunk 7]
```diff
@@ -20,6 +20,8 @@ per-file-ignores = __init__.py:F401 patch_saver_restore.py:E111,E114,
 # - W503 (no line breaks before binary operator): not PEP8-compliant; triggered by Black-formatted code
 # - C812-C816 (missing trailing comma): stylistic choice
 ignore = D1,D200,D202,D203,D205,D212,D4,E203,W503,C812,C813,C814,C815,C816
+# Disable flake8-import-restrictions
+extend-ignore = I20
```
Comment (wes-turner) [inline examples/tests/.flake8:24]: Same question as above, why does this work? Should it be `IMR2` or `IMR`?

**Review by wes-turner — APPROVED** [review]

File: harness/determined/pytorch/_pytorch_context.py [hunk 8]
```diff
@@ -4,7 +4,7 @@
 from typing import Any, Callable, Dict, Iterator, List, Optional, Set, Tuple, Type, Union
 
 import torch
-import torch.nn as nn
+from torch import nn
```
[the remark this answers is not in the document]
Comment (ioga) [reply harness/determined/pytorch/_pytorch_context.py:7]: semantically these end up being slightly different: 
- you can do `from determined.core import CheckpointContext`, i.e. import a non-module attribute in the `core` module
- but you can't do `import determined.core.CheckpointContext as CheckpointContext`.

for importing modules, it's the same. if we only allow importing modules, this does not matter. looks irrelevant for us.

torch docs use `import torch.nn as nn` quite often, probably because they are not bound by the same style guide.

File: e2e_tests/.flake8 (hunk 3 above)
Comment (ioga) [reply e2e_tests/.flake8:8]: it was on the ml sys agenda to discuss the plan for addressing these. I'd like to have this discussion before we can add concrete guidance here.

File: e2e_tests/tests/nightly/test_convergence.py (hunk 5 above)
Comment (ioga) [reply e2e_tests/tests/nightly/test_convergence.py:165]: I2000 is a different error, `Imports are only allowed on module level.`. which makes sense to enforce until the user is doing it intentionally.

File: e2e_tests/.flake8 (hunk 6 above)
Comment (ioga) [reply e2e_tests/.flake8:9]: the author renamed it in 2.0, while we're using [1.1.1](https://github.com/atollk/flake8-import-restrictions/tree/v1.1.1) because v2.0 needs flake8 5.x (we have 3.x), and does not support python 3.7.

File: examples/tests/.flake8 (hunk 7 above)
Comment (ioga) [reply examples/tests/.flake8:24]: same as above, 2.0 vs 1.1.1
